File: packages/wasmvm/wasmlib/src/assets.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use std::collections::{BTreeMap, HashSet};

use crate::*;

const HAS_BASE_TOKENS: u8 = 0x80;
const HAS_NATIVE_TOKENS: u8 = 0x40;
const HAS_NFTS: u8 = 0x20;

#[derive(Clone)]
pub struct ScAssets {
    base_tokens: u64,
    native_tokens: BTreeMap<Vec<u8>, ScBigInt>,
    nfts: HashSet<ScNftID>,
}

impl ScAssets {
    pub fn new(buf: &[u8]) -> ScAssets {
        let mut assets = ScAssets::new_base_tokens(0);
        if buf.len() == 0 {
            return assets;
        }

        let mut dec = WasmDecoder::new(buf);
        let flags = uint8_decode(&mut dec);
        if flags == 0x00 {
            return assets;
        }
        if (flags & HAS_BASE_TOKENS) != 0 {
            assets.base_tokens = dec.vlu_decode(64);
        }
        if (flags & HAS_NATIVE_TOKENS) != 0 {
            let size = dec.vlu_decode(16);
            for _i in 0..size {
                let token_id = token_id_decode(&mut dec);
                let amount = big_int_decode(&mut dec);
                assets.native_tokens.insert(token_id.to_bytes(), amount);
            }
        }
        if (flags & HAS_NFTS) != 0 {
            let size = dec.vlu_decode(16);
            for _i in 0..size {
                let nft_id = nft_id_decode(&mut dec);
                assets.nfts.insert(nft_id);
            }
        }
        assets
    }

    pub fn new_base_tokens(base_token_num: u64) -> ScAssets {
        return ScAssets {
            base_tokens: base_token_num,
            native_tokens: BTreeMap::new(),
            nfts: HashSet::new(),
        };
    }

    pub fn balances(&self) -> ScBalances {
        ScBalances {
            assets: self.clone(),
        }
    }

    pub fn is_empty(&self) -> bool {
        if self.base_tokens != 0 {
            return false;
        }
        for (_key, val) in self.native_tokens.iter() {
            if !val.is_zero() {
                return false;
            }
        }
        self.nfts.len() == 0
    }

    pub fn nft_ids(&self) -> Vec<ScNftID> {
        let mut nfts: Vec<ScNftID> = Vec::new();
        for nft in &self.nfts {
            nfts.push(nft.clone());
        }
        nfts.sort();
        nfts
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        let mut enc = WasmEncoder::new();
        if self.is_empty() {
            return vec![0; 1];
        }

        let mut flags = 0 as u8;
        if self.base_tokens != 0 {
            flags |= HAS_BASE_TOKENS;
        }
        if self.native_tokens.len() != 0 {
            flags |= HAS_NATIVE_TOKENS;
        }
        if self.nfts.len() != 0 {
            flags |= HAS_NFTS;
        }
        uint8_encode(&mut enc, flags);
        
        if (flags & HAS_BASE_TOKENS) != 0 {
            enc.vlu_encode(self.base_tokens);
        }
        if (flags & HAS_NATIVE_TOKENS) != 0 {
            enc.vlu_encode(self.native_tokens.len() as u64);
            for token_id in self.token_ids() {
                token_id_encode(&mut enc, &token_id);
                let amount = self.native_tokens.get(&token_id.to_bytes());
                big_int_encode(&mut enc, amount.unwrap());
            }
        }
        if (flags & HAS_NFTS) != 0 {
            enc.vlu_encode(self.nfts.len() as u64);
            for nft_id in self.nft_ids() {
                nft_id_encode(&mut enc, &nft_id);
            }
        }
        enc.buf()
    }

    pub fn token_ids(&self) -> Vec<ScTokenID> {
        let mut tokens: Vec<ScTokenID> = Vec::new();
        for (key, _val) in &self.native_tokens {
            tokens.push(token_id_from_bytes(key));
        }
        tokens.sort();
        tokens
    }
}

#[derive(Clone)]
pub struct ScBalances {
    assets: ScAssets,
}

impl ScBalances {
    fn new() -> ScBalances {
        ScBalances {
            assets: ScAssets::new(&[]),
        }
    }

    pub fn balance(&self, token: &ScTokenID) -> ScBigInt {
        let key = token.to_bytes();
        if !self.assets.native_tokens.contains_key(&key) {
            return ScBigInt::new();
        }
        self.assets.native_tokens.get(&key).unwrap().clone()
    }

    pub fn base_tokens(&self) -> u64 {
        self.assets.base_tokens
    }

    pub fn is_empty(&self) -> bool {
        self.assets.is_empty()
    }

    pub fn nft_ids(&self) -> Vec<ScNftID> {
        let mut nft_ids = Vec::new();
        for nft_id in self.assets.nfts.iter() {
            nft_ids.push(*nft_id);
        }
        return nft_ids;
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        self.assets.to_bytes()
    }

    pub fn token_ids(&self) -> Vec<ScTokenID> {
        self.assets.token_ids()
    }
}

#[derive(Clone)]
pub struct ScTransfer {
    balances: ScBalances,
}

impl ScTransfer {
    pub fn new() -> ScTransfer {
        ScTransfer {
            balances: ScBalances::new(),
        }
    }

    pub fn from_balances(balances: &ScBalances) -> ScTransfer {
        let mut transfer = ScTransfer::base_tokens(balances.base_tokens());
        for token_id in balances.token_ids().iter() {
            transfer.set(token_id, &balances.balance(token_id))
        }
        for nft_id in balances.nft_ids().iter() {
            transfer.add_nft(nft_id);
        }
        transfer
    }

    pub fn base_tokens(amount: u64) -> ScTransfer {
        let mut transfer = ScTransfer::new();
        transfer.balances.assets.base_tokens = amount;
        transfer
    }

    pub fn nft(nft_id: &ScNftID) -> ScTransfer {
        let mut transfer = ScTransfer::new();
        transfer.add_nft(nft_id);
        transfer
    }

    pub fn tokens(token_id: &ScTokenID, amount: &ScBigInt) -> ScTransfer {
        let mut transfer = ScTransfer::new();
        transfer.set(token_id, amount);
        transfer
    }

    pub fn add_nft(&mut self, nft_id: &ScNftID) {
        self.balances.assets.nfts.insert(nft_id.clone());
    }

    pub fn set(&mut self, token: &ScTokenID, amount: &ScBigInt) {
        self.balances
            .assets
            .native_tokens
            .insert(token.to_bytes(), amount.clone());
    }
}

impl std::ops::Deref for ScTransfer {
    type Target = ScBalances;
    fn deref(&self) -> &Self::Target {
        &self.balances
    }
}

File: packages/wasmvm/wasmlib/as/wasmlib/assets.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {ScDict} from './dict';
import {ScTokenID, ScTokenIDLength, tokenIDDecode, tokenIDEncode, tokenIDFromBytes} from './wasmtypes/sctokenid';
import {ScUint64Length, uint64FromBytes, uint64ToBytes} from './wasmtypes/scuint64';
import {bigIntDecode, bigIntEncode, ScBigInt} from './wasmtypes/scbigint';
import {nftIDDecode, nftIDEncode, ScNftID} from './wasmtypes/scnftid';
import {WasmDecoder, WasmEncoder} from './wasmtypes/codec';
import {uint8Decode, uint8Encode} from "./wasmtypes/scuint8";

const hasBaseTokens: u8 = 0x80;
const hasNativeTokens: u8 = 0x40;
const hasNFTs: u8 = 0x20;

export class ScAssets {
    baseTokens: u64 = 0;
    nativeTokens: Map<string, ScBigInt> = new Map();
    nfts: Map<string, ScNftID> = new Map();

    public constructor(buf: Uint8Array | null) {
        if (buf === null || buf.length == 0) {
            return this;
        }

        const dec = new WasmDecoder(buf);
        const flags = uint8Decode(dec);
        if (flags == 0x00) {
            return this;
        }

        if ((flags & hasBaseTokens) != 0) {
            this.baseTokens = dec.vluDecode(64);
        }
        if ((flags & hasNativeTokens) != 0) {
            let size = dec.vluDecode(16);
            for (; size > 0; size--) {
                const tokenID = tokenIDDecode(dec);
                const amount = bigIntDecode(dec);
                this.nativeTokens.set(ScDict.toKey(tokenID.id), amount);
            }
        }
        if ((flags & hasNFTs) != 0) {
            let size = dec.vluDecode(16);
            for (; size > 0; size--) {
                const nftID = nftIDDecode(dec);
                this.nfts.set(ScDict.toKey(nftID.id), nftID);
            }
        }
    }

    public balances(): ScBalances {
        return new ScBalances(this);
    }

    public isEmpty(): bool {
        if (this.baseTokens != 0) {
            return false;
        }
        const values = this.nativeTokens.values();
        for (let i = 0; i < values.length; i++) {
            if (!values[i].isZero()) {
                return false;
            }
        }
        return this.nfts.size == 0;
    }

    public nftIDs(): ScNftID[] {
        const nftIDs: ScNftID[] = [];
        const keys = this.nfts.keys().sort();
        for (let i = 0; i < keys.length; i++) {
            const nftID = this.nfts.get(keys[i]);
            nftIDs.push(nftID);
        }
        return nftIDs;
    }

    public toBytes(): Uint8Array {
        const enc = new WasmEncoder();
        if (this.isEmpty()) {
            return new Uint8Array(1);
        }

        let flags = 0x00 as u8;
        if (this.baseTokens != 0) {
            flags |= hasBaseTokens;
        }
        if (this.nativeTokens.size != 0) {
            flags |= hasNativeTokens;
        }
        if (this.nfts.size != 0) {
            flags |= hasNFTs;
        }
        uint8Encode(enc, flags);

        if ((flags & hasBaseTokens) != 0) {
            enc.vluEncode(this.baseTokens);
        }
        if ((flags & hasNativeTokens) != 0) {
            const keys = this.nativeTokens.keys().sort();
            enc.vluEncode(keys.length as u64);
            for (let i = 0; i < keys.length; i++) {
                const tokenID = ScDict.fromKey(keys[i]);
                enc.fixedBytes(tokenID, ScTokenIDLength);
                const amount = this.nativeTokens.get(keys[i]);
                bigIntEncode(enc, amount);
            }
        }
        if ((flags & hasNFTs) != 0) {
            const keys = this.nfts.keys().sort();
            enc.vluEncode(keys.length as u64);
            for (let i = 0; i < keys.length; i++) {
                const nftID = this.nfts.get(keys[i]);
                nftIDEncode(enc, nftID);
            }
        }
        return enc.buf();
    }

    public tokenIDs(): ScTokenID[] {
        const tokenIDs: ScTokenID[] = [];
        const keys = this.nativeTokens.keys().sort();
        for (let i = 0; i < keys.length; i++) {
            const keyBytes = ScDict.fromKey(keys[i]);
            const tokenID = tokenIDFromBytes(keyBytes);
            tokenIDs.push(tokenID);
        }
        return tokenIDs;
    }
}

export class ScBalances {
    assets: ScAssets;

    constructor(assets: ScAssets) {
        this.assets = assets;
    }

    public balance(tokenID: ScTokenID): ScBigInt {
        const mapKey = ScDict.toKey(tokenID.id);
        if (!this.assets.nativeTokens.has(mapKey)) {
            return new ScBigInt();
        }
        return this.assets.nativeTokens.get(mapKey);
    }

    public baseTokens(): u64 {
        return this.assets.baseTokens;
    }

    public isEmpty(): bool {
        return this.assets.isEmpty();
    }

    public nftIDs(): ScNftID[] {
        return this.assets.nftIDs();
    }

    public toBytes(): Uint8Array {
        return this.assets.toBytes();
    }

    public tokenIDs(): ScTokenID[] {
        return this.assets.tokenIDs();
    }
}

export class ScTransfer extends ScBalances {
    public constructor() {
        super(new ScAssets(null));
    }

    public static fromBalances(balances: ScBalances): ScTransfer {
        const transfer = ScTransfer.baseTokens(balances.baseTokens());
        const tokenIDs = balances.tokenIDs();
        for (let i = 0; i < tokenIDs.length; i++) {
            const tokenID = tokenIDs[i];
            transfer.set(tokenID, balances.balance(tokenID));
        }
        const nftIDs = balances.nftIDs();
        for (let i = 0; i < nftIDs.length; i++) {
            transfer.addNFT(nftIDs[i]);
        }
        return transfer;
    }

    public static baseTokens(amount: u64): ScTransfer {
        const transfer = new ScTransfer();
        transfer.assets.baseTokens = amount;
        return transfer;
    }

    public static nft(nftID: ScNftID): ScTransfer {
        const transfer = new ScTransfer();
        transfer.addNFT(nftID);
        return transfer;
    }

    public static tokens(tokenID: ScTokenID, amount: ScBigInt): ScTransfer {
        const transfer = new ScTransfer();
        transfer.set(tokenID, amount);
        return transfer;
    }

    public addNFT(nftID: ScNftID): void {
        this.assets.nfts.set(ScDict.toKey(nftID.id), nftID);
    }

    public set(tokenID: ScTokenID, amount: ScBigInt): void {
        const mapKey = ScDict.toKey(tokenID.id);
        this.assets.nativeTokens.set(mapKey, amount);
    }
}
